text/0000-inline-asm.md

```diff
@@ -0,0 +1,869 @@
+writing position independent code impossible.
+
+As a stop-gap the LDC implementation of D provides a `llvmasm` feature that binds it closely
+to LLVM IR's inline assembly.
+
+We believe it would be unfortunate to put Rust into a similar situation, making certain
+architectures a second-class citizen with respect to inline assembly.
+
+## Provide intrinsics for each instruction
+
+In discussions it is often postulated that providing intrinsics is a better solution to the
+problems at hand.
+However, particularly where precise timing, and full control over the number of generated
+instructions is required intrinsics fall short.
+
+Intrinsics are of course still useful and have their place for inserting specific instructions.
+E.g. making sure a loop uses vector instructions, rather than relying on auto-vectorization.
+
+However, inline assembly is specifically designed for cases where more control is required.
+Also providing an intrinsic for every (potentially obscure) instruction that is needed
+e.g. during early system boot in kernel code is unlikely to scale.
+
+## Make the `asm!` macro return outputs
+
+It has been suggested that the `asm!` macro could return its outputs like the LLVM statement does.
+The benefit is that it is clearer to see that variables are being modified.
+Particular in the case of initialization it becomes more obvious what is happening.
+On the other hand by necessity this splits the direction and constraint specification from
+the variable name, which makes this syntax overall harder to read.
+
+```rust
+fn mul(a: u32, b: u32) -> u64 {
+    let (lo, hi) = unsafe {
+        asm!("mul {}", in(reg) a, in("eax") b, lateout("eax"), lateout("edx"))
+    };
+
+    hi as u64 << 32 + lo as u64
+}
+```
+
+# Prior art
+[prior-art]: #prior-art
+
+## GCC inline assembly
+
+The proposed syntax is very similar to GCC's inline assembly in that it is based on string substitution while leaving actual interpretation of the final string to the assembler. However GCC uses poorly documented single-letter constraint codes and template modifiers. Clang tries to emulate GCC's behavior, but there are still several cases where its behavior differs from GCC's.
+
+The main reason why this is so complicated is that GCC's inline assembly basically exports the raw internals of GCC's register allocator. This has resulted in many internal constraint codes and modifiers being widely used, despite them being completely undocumented.
+
+## D & MSVC inline assembly
+
+See the section [above][dsl].
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+None
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Flag outputs
+
+GCC supports a special type of output which allows an asm block to return a `bool` encoded in the condition flags register. This allows the compiler to branch directly on the condition flag instead of materializing the condition as a `bool`.
```

Comment:
This would be super useful for Intel ADCX/ADOX instructions for multi-precision integer math. I have an example of a simple assembly multiplication routine [here](https://github.com/0xProject/OpenZKP/blob/6e59a51e39eaa33331d0860b6b60d7a70a19f709/algebra/u256/src/algorithms/assembly.rs#L86) It would be useful for me if I could specify the carry and overflow flags as inputs/outputs.

This is in line with what the LLVM instrinsics for these instructions do. See [`core::arch::x86_64::_addcarryx_u64`](https://doc.rust-lang.org/core/arch/x86_64/fn._addcarryx_u64.html).

Unfortunately, again it seems like LLVM just ignores it. As mentioned in https://github.com/rust-lang/stdarch/issues/666#issuecomment-485339414, there is an open LLVM issue and it does not seem easy:

> The X86 backend isn't currently set up to model the C flag and O flag separately. We model all of the flags as one register. Because of this we can't interleave the flag dependencies. We would need to do something about that before it makes sense to implement _addcarryx_u64 as anything other than plain adc.
> -- https://bugs.llvm.org/show_bug.cgi?id=41546#c1